docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,192 @@
+case labels match. The `default` case is evaluated last, regardless
+of its textual order.
+
+## `when` clauses in switch expressions
+
+You can make special cases for those shapes that have 0 area by using
+a `when` clause on the switch label. A square with side length of 0, or
+a circle with a radius of 0 has a 0 area. You specify that condition
+using a `when` clause on the switch label:  
+
+[!code-csharp[ComputeDegenerateShapes](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#07_ComputeDegenerateShapes "Compute shapes with 0 area")]
+
+This change demonstrates a few important points on the new syntax. First,
+multiple switch labels can be applied to one switch section. The statement
+block is executed when any of those labels is true. In this instance,
+if the switch expression is either a circle or a square with 0 area, the
+method returns the constant 0.
+
+This example introduces two different variables in the two switch labels
+for the first switch block. Notice that the statements in this switch block
+do not use either the variables `c` (for the circle), or `s` (for the square).
+Neither of those variables is definitely assigned in this switch block.
+Clearly one has been assigned, but static analysis cannot know which one. For that reason,
+most times when you use multiple switch labels for the same block, you won't
+introduce a new variable in the `case` statement, or you will only use the
+variable only in the `when` clause.
+
+Having adding those degenarate shapes, let's add a couple more shape types:
+a rectangle and a triangle:
+
+[!code-csharp[RectangleAndTriangle](../../samples/csharp/PatternMatching/PatternMatching/Shapes.cs#08_RectangleAndTriangle "Rectangle and Triangle definition")]
+
+Next, you'll add switch labels and blocks for those shapes in the `ComputeArea` method:
+
+[!code-csharp[AddRectangleAndTriangle](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#09_AddRectangleAndTriangle "Add rectangle and triangle")]
+
+ This set of changes adds switch labels for the degenerate case, and labels
+ and blocks for each of the new shapes. 
+
+Finally, you can add a `null` case to ensure the argument is not `null`:
+
+[!code-csharp[NullCase](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#10_NullCase "Add null case")]
+
+The special case for the `null` pattern is interesting because the constant `null` does
+not have a type, but can be converted to any reference type or nullable
+type. 
+
+## Conclusions
+
+*Pattern Matching expressions* enable you to easily manage control flow
+among different variables and types that are not related by an inheritance
+hierarchy. You can also the control logic to use any condition you test on
+the variable. It enables patterns and idioms that you'll need more often
+as we build more distributed applications, where data and the methods that
+manipulate that data are separate. You'll notice that the shape structs
```

Comment:
"where data and the methods that manipulate that data are separate": this is very clear, and should probably appear earlier in the article, because it provides the main justification for the feature. The article is heavy on "how" in its present state, but relatively light on "why".